[SPARK-47509][SQL] Block subquery expressions in lambda and higher-order functions

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR blocks subquery expressions in lambda and higher-order functions, since these can generate incorrect results.

### Why are the changes needed?

It is safer to block these query patterns which can generate incorrect results for now. We can consider re-enabling these patterns later after fixing the correctness problem.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds unit test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

dtenedor committed:
commit

dtenedor commented:
cc @cloud-fan

jchen5 reviewed (commented):
This blocks a subquery that contains a lambda, but the problem case was actually the reverse, a lambda that contains a subquery, e.g. `SELECT arrcol, forall(arrcol, e -> e in (select 1)) as forallout FROM bug_test_table;`

jchen5 commented:
CC @agubichev also

dongjoon-hyun reviewed (commented):
Thank you for making a PR, @dtenedor .

According to the reported JIRA Affected Version, is this Apache Spark 4.0.0 only correctness bug? Do you happen to know when this correctness bug is introduced?

![Screenshot 2024-03-21 at 18 22 35](https://github.com/apache/spark/assets/9700541/865d0dab-0735-44b5-b4ee-fdd9b226b723)

dtenedor commented:
> According to the reported JIRA Affected Version, is this Apache Spark 4.0.0 only correctness bug? Do you happen to know when this correctness bug is introduced?

@jchen5 do you know?

jchen5 commented:
I suspect it has existed as long as lambdas have. So at least like 5 years?

dtenedor committed:
switch to checking for subqueries with lambdas inside, instead of vice versa

jchen5 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:
```diff
@@ -138,4 +141,16 @@ object ResolveLambdaVariables extends Rule[LogicalPlan] {
   }
+
+  /**
+   * SPARK-47509: There is a correctness bug when lambdas or higher-order functions appear within
```
Comment should be updated too: appear within -> contain

dtenedor commented:
@jchen5 @dongjoon-hyun I updated it to switch to checking for subqueries with lambdas inside, instead of vice versa, per request. Please take another look :)

jchen5 reviewed (approved):
Thanks for the fix!

dtenedor committed:
respond to code review comments

cloud-fan commented on common/utils/src/main/resources/error/error-classes.json:
```diff
@@ -4457,6 +4457,11 @@
           "Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses: <sqlExprs>."
         ]
       },
+      "LAMBDA_OR_HIGHER_ORDER_FUNCTION" : {
```
hmmm, are these two the same thing?

cloud-fan replied:
Reading the code comment
```
/**
 * A higher order function takes one or more (lambda) functions and applies these to some objects.
 * The function produces a number of variables which can be consumed by some lambda function.
 */
```

I think for use-facing error message, we can just say "higher order functions". For the code change, we can check both `HigherOrderFunction` and `LambdaFunction` to  be safe.

dtenedor replied:
Sounds good, this is simpler. Done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:
```diff
@@ -138,4 +141,16 @@ object ResolveLambdaVariables extends Rule[LogicalPlan] {
+  private def checkForSubqueryExpressions(expression: Expression): Unit = {
+    if (expression.containsPattern(PLAN_EXPRESSION) &&
+      !conf.getConf(SQLConf.ALLOW_SUBQUERY_EXPRESSIONS_IN_LAMBDAS_AND_HIGHER_ORDER_FUNCTIONS)) {
+      throw QueryCompilationErrors.subqueryExpressionInLambdaOrHigherOrderFunctionNotAllowedError()
```
shall we update the PR title to match the new code?

dtenedor replied:
This is done.

dtenedor committed:
sync

dtenedor committed:
regenerate error classes doc

dtenedor commented:
Thanks @cloud-fan for your review! Responded to your suggestions and updated the error classes documentation and fixed the conflict with the `master` branch, please take another look.

cloud-fan commented on common/utils/src/main/resources/error/error-classes.json:
```diff
@@ -4496,6 +4496,11 @@
           "Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses: <sqlExprs>."
         ]
       },
+      "LAMBDA_OR_HIGHER_ORDER_FUNCTION" : {
```
```suggestion
      "HIGHER_ORDER_FUNCTION" : {
```

cloud-fan commented on docs/sql-error-conditions-unsupported-subquery-expression-category-error-class.md:
```diff
@@ -50,6 +50,10 @@ A correlated outer name reference within a subquery expression body was not foun
 
 Expressions referencing the outer query are not supported outside of WHERE/HAVING clauses: `<sqlExprs>`.
 
+## LAMBDA_OR_HIGHER_ORDER_FUNCTION
```
```suggestion
## HIGHER_ORDER_FUNCTION
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -620,6 +620,12 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
 
+  def subqueryExpressionInLambdaOrHigherOrderFunctionNotAllowedError(): Throwable = {
+    new AnalysisException(
+      errorClass = "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.LAMBDA_OR_HIGHER_ORDER_FUNCTION",
```
```suggestion
      errorClass = "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.HIGHER_ORDER_FUNCTION",
```

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ResolveSubquerySuite.scala:
```diff
@@ -270,4 +271,32 @@ class ResolveSubquerySuite extends AnalysisTest {
+    assertAnalysisErrorClass(
+      inputPlan = lambdaPlanScanFromTable,
+      expectedErrorClass =
+        "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.LAMBDA_OR_HIGHER_ORDER_FUNCTION",
```
```suggestion
        "UNSUPPORTED_SUBQUERY_EXPRESSION_CATEGORY.HIGHER_ORDER_FUNCTION",
```

cloud-fan committed:
Apply suggestions from code review

cloud-fan commented:
thanks, merging to master!
